## Practical implementation of Decision Tree Algorithm
Submitted by:- Ambarish Singh

### Importing Necessary Library

In [56]:
## Importing Necessary Library
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV


### Importing CSV File

In [4]:
## Importing CSV File
df = pd.read_csv("https://raw.githubusercontent.com/shrikant-temburwar/Wine-Quality-Dataset/master/winequality-red.csv",sep=";")

In [5]:
## Checking Top 5 Rows
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [6]:
## Checking Bottom 5 Rows
df.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6

In [7]:
## Checking Shape of the Dataset, here there are total 1599 Rows and 12 Columns.
df.shape

(1599, 12)

In [8]:
## Checking Target Column Unique Details.
df['quality'].unique()

array([5, 6, 7, 4, 8, 3], dtype=int64)

In [10]:
## Checking Length of target Columns Unique Details.
len(df['quality'].unique())

6

In [9]:
## Checking Total Count of Target Columns Unique Details
df['quality'].value_counts()


5    681
6    638
7    199
4     53
8     18
3     10
Name: quality, dtype: int64

In [11]:
## Checking Basic Statistics Analysis via Transpose 
df.describe().T 


count       mean        std      min      25%  \
fixed acidity         1599.0   8.319637   1.741096  4.60000   7.1000   
volatile acidity      1599.0   0.527821   0.179060  0.12000   0.3900   
citric acid           1599.0   0.270976   0.194801  0.00000   0.0900   
residual sugar        1599.0   2.538806   1.409928  0.90000   1.9000   
chlorides             1599.0   0.087467   0.047065  0.01200   0.0700   
free sulfur dioxide   1599.0  15.874922  10.460157  1.00000   7.0000   
total sulfur dioxide  1599.0  46.467792  32.895324  6.00000  22.0000   
density               1599.0   0.996747   0.001887  0.99007   0.9956   
pH                    1599.0   3.311113   0.154386  2.74000   3.2100   
sulphates             1599.0   0.658149   0.169507  0.33000   0.5500   
alcohol               1599.0  10.422983   1.065668  8.40000   9.5000   
quality               1599.0   5.636023   0.807569  3.00000   5.0000   

                           50%        75%        max  
fixed acidity          7.90000   9.200000   15.90000  
volatile acidity       0.52000   0.640000    1.58000  
citric acid            0.26000   0.420000    1.00000  
residual sugar         2.20000   2.600000   15.50000  
chlorides              0.07900   0.090000    0.61100  
free sulfur dioxide   14.00000  21.000000   72.00000  
total sulfur dioxide  38.00000  62.000000  289.00000  
density                0.99675   0.997835    1.00369  
pH                     3.31000   3.400000    4.01000  
sulphates              0.62000   0.730000    2.00000  
alcohol               10.20000  11.100000   14.90000  
quality                6.00000   6.000000    8.00000

In [12]:
## Checking Duplicate Rows in Dataset and there are total 240 Duplicate rows in dataset.
df.duplicated().sum()    


240

In [13]:
## Droping Duplicate Rows from dataset.
df = df.drop_duplicates()  


In [14]:
## Again Checking Duplicate Rows in dataset, Now Dupliacte Row is Zero after Droping Duplicated rows.
df.duplicated().sum()


0

### Creating Independent Feature

In [16]:
X = df.drop("quality", axis =1)

### Creating Dependent Feature

In [17]:
y = df['quality']

# Model Building

In [18]:
from sklearn.model_selection import train_test_split,GridSearchCV

In [36]:
X_train,X_test,y_train,y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [37]:
## Scaling is not Required in Decision tree Problem

'''
from sklearn.preprocessing import StandardScaler
sclr = StandardScaler()
sclr.fit()
sclr.transform()
'''

'\nfrom sklearn.preprocessing import StandardScaler\nsclr = StandardScaler()\nsclr.fit()\nsclr.transform()\n'

# Using Decision Tree classifier Algorithm

In [38]:
from sklearn.tree import DecisionTreeClassifier
model  = DecisionTreeClassifier()

In [39]:
## Model Fitting with Decision Tree Classifier on Training dataset
model.fit(X_train,y_train)

DecisionTreeClassifier()

In [40]:
model.score(X_train,y_train)

1.0

In [41]:
## Model Prediction on Testing Dataset
y_predict = model.predict(X_test)

In [42]:
from sklearn.metrics import accuracy_score

In [43]:
## Accuracy of the Model using Decision Tree Classifier Algorithm
accuracy_score(y_test,y_predict)


0.5073529411764706

**Observations**

* Getting 51% approx Model Accuracy using Decision Tree classifier Algorithm

In [44]:
## Try to check with diff.. diff.. Algorithm
### Logistic Regression, SVR,etc
## Going forward [DT,AB,XG]
## EXtra :- Do Feature Engineering Details by your own on above Dataset.

## Applying GridSearchCV Hyperparameter tuning

In [46]:
## Applying Different different Parameter Using gridSearchCV

grid_param = {
    'criterion': ['gini', 'entropy'],
    'max_depth' : range(2,32,1),
    'min_samples_leaf' : range(1,10,1),
    'min_samples_split': range(2,10,1),
    'splitter' : ['best', 'random']
    
}

In [47]:
## Importing GridSearchCV Library

from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(estimator= model, param_grid = grid_param, cv = 5)


In [49]:
## GridSearchCV take a lot of time to RUN.
grid_search.fit(X_train,y_train)   


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': range(2, 32),
                         'min_samples_leaf': range(1, 10),
                         'min_samples_split': range(2, 10),
                         'splitter': ['best', 'random']})

In [50]:
## Finding Best param Model using GridSearch
grid_search.best_params_


{'criterion': 'gini',
 'max_depth': 4,
 'min_samples_leaf': 6,
 'min_samples_split': 7,
 'splitter': 'random'}

In [52]:
## Applying Best Param Model.
model_with_best_params = DecisionTreeClassifier( criterion = 'gini',
                                                 max_depth = 4,
                                                 min_samples_leaf = 6,
                                                 min_samples_split = 7,
                                                 splitter = 'random')


In [53]:
## Fitting Best Param Model to Training Dataset.
model_with_best_params.fit(X_train,y_train)


DecisionTreeClassifier(max_depth=4, min_samples_leaf=6, min_samples_split=7,
                       splitter='random')

In [54]:
## Applying Prediction with Best Param Model
y_prediction2 = model_with_best_params.predict(X_test)


In [55]:
## Checking Accuracy of Model After Doing HyperParameter Tuning using GridSearchCV 
accuracy_score(y_test, y_prediction2)


0.571078431372549

**Observations**

* Current Accuracy is Approx. 57%  (After doing Hyperparameter Tuning )
* Previous Accuracy is Approx. 51% ( Before hyperparameter Tuning )


**Summary**

* After Doing Hyperparameter Tuning Approx. 6% Accuracy improve.

## Thank You